src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/SqlClientDiagnostic.cs

```diff
@@ -0,0 +1,676 @@
+using System.Data;
+
+namespace Microsoft.Data.SqlClient.Diagnostics
+{
+    /// <include file='../../../../../../../doc/snippets/Microsoft.Data.SqlClient/SqlClientDiagnostic.xml' path='docs/members[@name="SqlClientDiagnostic"]/SqlClientDiagnostic/*'/>
+    public abstract class SqlClientDiagnostic : IReadOnlyList<KeyValuePair<string, object>>
+    {
+        /// <include file='../../../../../../../doc/snippets/Microsoft.Data.SqlClient/SqlClientDiagnostic.xml' path='docs/members[@name="SqlClientDiagnostic"]/CommonPropertyCount/*'/>
+        protected const int CommonPropertyCount = 3;
+
+        /// <include file='../../../../../../../doc/snippets/Microsoft.Data.SqlClient/SqlClientDiagnostic.xml' path='docs/members[@name="SqlClientDiagnostic"]/ctor/*'/>
+        protected SqlClientDiagnostic(Guid operationId, string operation, long timestamp)
+        {
+            OperationId = operationId;
+            Operation = operation;
+            Timestamp = timestamp;
+        }
+
+        /// <include file='../../../../../../../doc/snippets/Microsoft.Data.SqlClient/SqlClientDiagnostic.xml' path='docs/members[@name="SqlClientDiagnostic"]/OperationId/*'/>
+        public Guid OperationId { get; }
+        /// <include file='../../../../../../../doc/snippets/Microsoft.Data.SqlClient/SqlClientDiagnostic.xml' path='docs/members[@name="SqlClientDiagnostic"]/Operation/*'/>
+        public string Operation { get; }
+        /// <include file='../../../../../../../doc/snippets/Microsoft.Data.SqlClient/SqlClientDiagnostic.xml' path='docs/members[@name="SqlClientDiagnostic"]/Timestamp/*'/>
+        public long Timestamp { get; }
+
+        /// <inheritdoc/>>
+        public int Count => CommonPropertyCount + GetDerivedCount();
+
+        /// <inheritdoc/>>
+        public KeyValuePair<string, object> this[int index]
+        {
+            get
+            {
+                if (TryGetCommonProperty(index, out KeyValuePair<string, object> commonProperty))
+                {
+                    return commonProperty;
+                }
+                else
+                {
+                    return GetDerivedProperty(index - CommonPropertyCount);
+                }
+            }
+        }
+
+        /// <inheritdoc/>>
+        public IEnumerator<KeyValuePair<string, object>> GetEnumerator()
+        {
+            int count = Count;
+            for (var index = 0; index < count; index++)
+            {
+                yield return this[index];
+            }
+        }
+
+        /// <inheritdoc/>>
+        IEnumerator IEnumerable.GetEnumerator() => GetEnumerator();
+
+        /// <include file='../../../../../../../doc/snippets/Microsoft.Data.SqlClient/SqlClientDiagnostic.xml' path='docs/members[@name="SqlClientDiagnostic"]/TryGetCommonProperty/*'/>
+        protected bool TryGetCommonProperty(int index, out KeyValuePair<string, object> property)
+        {
+            switch (index)
+            {
+                case 0:
```

Comment:
I saw these index numbers are being used a lot in this class, would it be possible to pull them up as constant fields with more descriptive names just for code readability purpose?

Reply:
I tried that while I was working on the first class I implemented and found that it hurt readability more than it helped.

Consider either:
```csharp
	private const int ConnectionIdIndex = 0;
	private const int TransactionIdIndex = 1;
	private const int CommandIdIndex = 2

        protected sealed override KeyValuePair<string, object> GetDerivedProperty(int index) => index switch
        {
            ConnectionIdIndex => new KeyValuePair<string, object>(nameof(ConnectionId), ConnectionId),
            TransactionIdIndex => new KeyValuePair<string, object>(nameof(TransactionId), TransactionId),
            CommandIdIndex => new KeyValuePair<string, object>(nameof(Command), Command),
            _ => throw new IndexOutOfRangeException(nameof(index)),
        };
```

or:

```csharp
        protected sealed override KeyValuePair<string, object> GetDerivedProperty(int index) => index switch
        {
	    const int ConnectionIdIndex = 0;
	    const int TransactionIdIndex = 1;
	    const int CommandIdIndex = 2

            ConnectionIdIndex => new KeyValuePair<string, object>(nameof(ConnectionId), ConnectionId),
            TransactionIdIndex => new KeyValuePair<string, object>(nameof(TransactionId), TransactionId),
            CommandIdIndex => new KeyValuePair<string, object>(nameof(Command), Command),
            _ => throw new IndexOutOfRangeException(nameof(index)),
        };
```

Both are much longer and require you to mentally match up the constants by name when their only realy importantce is that they're sequential. They're also used in only on place (because each class has different properties) so the constants are only ever used once. I settled with just using the numbers directly because it was shorter and overall clearer.

For the same reasons I'm not really totally happy with the Name constants because I feel they make things less readable. In the case of the name the idea of having a unique required string that people will need to copy paste to use the individual class but that string not being defined in the class felt more wrong than the additional verbosity of having it locally.